### DISCUSSION TITLE
How do I use sktime to forecast categorical labels or classification time series using the following data

### DISCUSSION THREAD

**Discussion by xxl4tomxu98** [alan-turing-institute/sktime#3976] [Ideas]
I have data stream coming in continuously on daily basis, it is panel like structure like this:
![image](https://user-images.githubusercontent.com/62292177/209034884-bbfc5461-16fd-4284-9599-50ccde28e4b2.png)
I am trying to predict the label(last column, "ready for test") for each student continuously based on historical data. For instance, I want to determine when in the next few days(window) a student would be NOT ready for test and so on. Which is best way to formulate this problem for sktime? Time series forecasting? or Classification of a time series for each student?  Thanks

**Comment by fkiraly** [comment]
To diagnose the task:

it could be two things, the key question is: do we assume that, for a student, progression of time reduces the *uncertainty* about a certain label, e.g., are they either ready or not for the test all along, and we just gain more and more information about their readiness? Or, is the actual label also changing over time, i.e., they are initially not ready and they become increasingly ready?

If there is one "true" label per student and what changes is uncertainty:
this would be early time series classification.

If the "true" label varies with time and is observed with certainty:
this is supervised annotation with a categorical label, in a temporal on-line setting

**Reply by fkiraly** [reply]
for what we have in `sktime`:
* we have a few early classifiers, you can use `all_estimators` with the scitype `early-classifier` to look them up
* for supervised annotation with a categorical label, in a temporal on-line setting, I am not sure whether we have an dedicated functionality. @miraep8, @KatieBuc, @lmmentel might know whether one of the annotators is suitable for that.
* what you could also do is reduce/transform your problem into a time series classification problem, there are a lot of time series classifiers and feature extraction and pipeline building utilities in `sktime`. You can do this by considering, as samples, the time series of days, homework, breakfast, up until the label as the feature input, and the label on that day as the feature output. You would have to make sure that data from a given student ends up only in the train set, or the test set, but not both.

The last option would end up with the following X/y, for the first student:

feature-label pair 1: student 1, day 1 (time series of length 1); label = yes
feature-label pair 2: student 1, day 1 and 2 (time series of length 2); label = yes
feature-label pair 3: student 1, day 1 to 3 (time series of length 3); label = no
feature-label pair 4: student 1, day 1 to 4 (time series of length 4); label = maybe

note, this ends up with time series of unequal length, so you will need to find or build a classifier that does that.
The relevant tags are `"capability:unequal_length"` for classifieres and transformers, and `"capability:unequal_length:removes"` for transformers that change unequal length series to equal length (e.g., padding with zeros).

**Reply by fkiraly** [reply]
a baseline to compare to, of course, would be applying a tabular classifier directly to the table you have. This will not use the "history" of the student but only the data on the same day, but it is simpler than the above and should be a point of comparison.

**Comment by xxl4tomxu98** [comment]
Yeah, Franz, thanks for the rather comprehensive advice. I did do this base
line training with bunch of students using their everyday data without
history and use another bunch of students as testing set, but the scores
are pretty bad with accuracy in the 60% and ROC score in the same range.
That was why I came to you!

Thinking about this I started to judge it as perhaps annotations with a
time series problem, because everyday there will be actual data and label
coming in and it changes day by day. All I am interested to know is perhaps
what will happen in a longer time horizon so I can deal with it and try to
change this earlier prediction. While I am projecting into this longer
horizon, records are kept coming in everyday with actual label. These
actual labels change because of my intervention or someone else’s

If you think this is indeed annotation with time series, then I do need to
find tools available with help of your colleagues. Thanks

Tom

On Fri, Dec 23, 2022 at 6:38 AM Franz Király ***@***.***>
wrote:

> a baseline to compare to, of course, would be applying a tabular
> classifier directly to the table you have. This will not use the "history"
> of the student but only the data on the same day, but it is simpler than
> the above and should be a point of comparison.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sktime/sktime/discussions/3976#discussioncomment-4484119>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO3IBUMMSD5IXQWKF7EAHSTWOWFKBANCNFSM6AAAAAATGGO7G4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Tom Xu

**Reply by xxl4tomxu98** [reply]
@fkiraly, @miraep8, @KatieBuc, @lmmentel , I thought about it again, I am not sure what "supervised annotation with a categorical label, in a temporal on-line setting" actually means. But I know I do need to forecast the label for test readiness in projected forecast window, say 50 days, etc. This is to assume that if no intervention is acted upon the student at forecasting date. As date moves closer to the previously forecasted classification, I am already doing early classifications on more future dates. I hope the logic makes sense. But still not very sure about what is the best approach for it. 

In particular this case, if I do "capability:unequal_length" or "capability:unequal_length:removes" with "early_classifier", I end up with empty array "[]" for both cases, indicating no suitable estimator available, while if I just require "early_classifier", I got 1 classifier, "[('TEASER', sktime.classification.early_classification._teaser.TEASER)]". So I would run into problems trying to use early classification approach?

**Comment by xxl4tomxu98** [comment]
@miraep8, @KatieBuc, @lmmentel, Please let me know if any annotator estimators would fit this problem. Thanks

**Comment by lmmentel** [comment]
Hey @xxl4tomxu98, I not sure what kind of approach @fkiraly had in mind so it's hard for me to follow up on his comments. I can say that in the annotation subpackage in sktime we do currently have a few estimators implemented for unsupervised time series segmentation and anomaly detection but we do not have strictly supervised segmentation estimators implemented yet. From your description I don't think any of them would match your problem. If you want to learn more about supervised change point detection you check [this talk](https://learn.microsoft.com/en-us/events/user-international-r-user-conferences-user-international-r-user-2017-conference/introduction-to-optimal-changepoint-detection-algorithms-ii) by Toby Dylan Hocking but I guess the connection to your problem might not be obvious. 

Before jumping to more complicated solution I would probably explore simplest model that might works in this case. That is probably going back to methods that work on tabular data rather than sequences and try to extract some features from your sequences and classify them with one of the sklearn's standard classifiers. Best of luck!

**Reply by xxl4tomxu98** [reply]
Thank you @lmmentel , very much appreciated. I like @fkiraly 's suggestion on using reducer/transformer to turn each tabular row into a timer series as this:

feature-label pair 1: student 1, day 1 (time series of length 1); label = yes
feature-label pair 2: student 1, day 1 and 2 (time series of length 2); label = yes
feature-label pair 3: student 1, day 1 to 3 (time series of length 3); label = no
feature-label pair 4: student 1, day 1 to 4 (time series of length 4); label = maybe

Do you know any sktime transformer or reducer that does this transformation? e.g. turn each cell into a time series starting from row 0 and ending current row? Thanks
